libs/wingsdk/src/core/resource.ts

```diff
@@ -353,3 +341,69 @@ export class Display {
     this.hidden = props?.hidden;
   }
 }
+
+/**
+ * Binds an object (either data or resource) to a host.
+ *
+ * - Primitives and Duration objects are ignored.
+ * - Arrays, sets and maps and structs (Objects) are recursively bound.
+ * - Resources are bound to the host by calling their _bind() method.
+ *
+ * @param obj The object to bind.
+ * @param host The host to bind to
+ * @param ops The set of operations that may access the object.
+ */
+function bindObject(obj: any, host: IResource, ops: string[] = ["?"]): void {
+  switch (typeof obj) {
+    case "string":
+    case "boolean":
+    case "number":
+      return;
+
+    case "object":
+      if (Array.isArray(obj)) {
+        obj.forEach((item) => bindObject(item, host, ops));
+        return;
+      }
+
+      if (obj instanceof Duration) {
+        return;
+      }
+
+      if (obj instanceof Set) {
+        return Array.from(obj).forEach((item) => bindObject(item, host, ops));
+      }
+
+      if (obj instanceof Map) {
+        Array.from(obj.values()).forEach((item) => bindObject(item, host, ops));
+        return;
+      }
+
+      // if the object is a resource (i.e. has a "_toInflight" method"), then we lift and add to the
+      // list of bindings.
+      if (typeof (obj as IResource)._bind === "function") {
+        (obj as IResource)._bind(host, ops);
+
+        // add connection metadata (use "?" to indicate that we don't know the operation)
+        for (const op of ops) {
+          Resource.addConnection({
+            from: host,
+            to: obj,
+            relationship: op,
+          });
+        }
+
+        return;
+      }
+
+      // structs are just plain objects
+      if (obj.constructor.name === "Object") {
+        Object.values(obj).forEach((item) => bindObject(item, host, ops));
```

Comment:
I feel like there's something we're not quite modeling here -- the list of operations passed to this function are specifically operations about `obj`. (Or at least, that was the original intent). If a struct contains a bucket and a key value store, we shouldn't mix up performing `put` on a `cloud.Bucket` with performing `put` on a `cloud.KeyValueStore`. Do you see what I mean?

Reply:
Yes, good point. In the `struct` case, `ops` is just going to be `["?"]` anyway, so I think we are "fine".

As discussed yesterday, we need to revisit how "ops" are determined/propagated anyway.